core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java

```diff
@@ -478,16 +912,335 @@ SortedRangeSet build()
                     current = current.span(next);
                 }
                 else {
-                    result.put(current.getLow(), current);
+                    result.add(current);
                     current = next;
                 }
             }
 
             if (current != null) {
-                result.put(current.getLow(), current);
+                result.add(current);
             }
 
-            return new SortedRangeSet(type, result);
+            boolean[] inclusive = new boolean[2 * result.size()];
+            BlockBuilder blockBuilder = type.createBlockBuilder(null, 2 * result.size());
+            for (int rangeIndex = 0; rangeIndex < result.size(); rangeIndex++) {
+                Range range = result.get(rangeIndex);
+                writeRange(type, blockBuilder, inclusive, rangeIndex, range);
+            }
+
+            return new SortedRangeSet(type, inclusive, blockBuilder);
+        }
+    }
+
+    private static void writeRange(Type type, BlockBuilder blockBuilder, boolean[] inclusive, int rangeIndex, Range range)
+    {
+        inclusive[2 * rangeIndex] = range.getLow().getBound() == Marker.Bound.EXACTLY;
+        inclusive[2 * rangeIndex + 1] = range.getHigh().getBound() == Marker.Bound.EXACTLY;
+        if (range.getLow().getValueBlock().isEmpty()) {
+            blockBuilder.appendNull();
+        }
+        else {
+            type.appendTo(range.getLow().getValueBlock().get(), 0, blockBuilder);
+        }
+        if (range.getHigh().getValueBlock().isEmpty()) {
+            blockBuilder.appendNull();
+        }
+        else {
+            type.appendTo(range.getHigh().getValueBlock().get(), 0, blockBuilder);
+        }
+    }
+
+    private static void writeRange(Type type, BlockBuilder blockBuilder, boolean[] inclusive, int rangeIndex, RangeView range)
+    {
+        inclusive[2 * rangeIndex] = range.lowInclusive;
+        inclusive[2 * rangeIndex + 1] = range.highInclusive;
+        type.appendTo(range.lowValueBlock, range.lowValuePosition, blockBuilder);
+        type.appendTo(range.highValueBlock, range.highValuePosition, blockBuilder);
+    }
+
+    private static class RangeView
```

Comment:
Are you certain that this class is thoroughly tested?

Reply:
of course! ... it's supposed to be implicitly tested. i found a regression & added a test to `TestSortedRangeSet`